Add all relevant testcases for Arm Ethos-U85

zingo opened the pull request:
Add separate tests for Ethos-U85 to all backend operator tests.
Updated ethos-u-vela version to support more operators.

Signed-off-by: Per Åstrand <[per.astrand@arm.com](mailto:per.astrand@arm.com)>
Signed-off-by: Tom Allsop <[tom.allsop@arm.com](mailto:tom.allsop@arm.com)>

per committed:
Add all relevant testcases for Ethos-U85

Add separate tests for Ethos-U85 to all backend operator tests.

Signed-off-by: Per Åstrand <per.astrand@arm.com>
Signed-off-by: Tom Allsop <tom.allsop@arm.com>

Co-authored-by: Tom Allsop <tom.allsop@arm.com>
Change-Id: I923372adec92d095f11cb78aa55b675b1be7f070

freddan80 committed:
Increase memory allocated for inputs in case semi hosting is used.

Change-Id: Iffbed1aea3583a6c725367db852d411cb1eb7783

zingo commented:
@pytorchbot label ciflow/trunk

digantdesai commented on backends/arm/test/models/test_mobilenet_v2_arm.py:
```diff
@@ -97,3 +97,18 @@ def test_mv2_u55_BI(self):
             .to_executorch()
         )
+
+    def test_mv2_u85_BI(self):
```
share with u55? And run it on corstone in the future?

zingo replied:
> share with u55? 

How do you mean? We already have a test_mv2_u55_BI() ?

> And run it on corstone in the future?

Maybe we are trying out different ways to run larger models like mv2 and others in a different way then the unit test setup. Lest see how it turns out :)

digantdesai replied:
Share the body of the test I mean, not a big deal TBH

digantdesai commented on examples/arm/setup.sh:
```diff
@@ -216,7 +216,7 @@ function setup_vela() {
         git clone https://review.mlplatform.org/ml/ethos-u/ethos-u-vela
         repo_dir="${root_dir}/ethos-u-vela"
-        base_rev=d362f5443f67b1e6213a9d8f124edff758efac96
+        base_rev=fe0eaa55c5ed319f78c01978f3b40eb11a9bcb38
```
In the summary can we add if there is any major change with this rev? Like one or two line would be helpful to understand the rationale.

per replied:
Yes, good idea!

digantdesai commented on backends/arm/test/ops/test_add.py:
```diff
@@ -130,7 +131,22 @@ def test_add_tosa_BI(self, test_data: torch.Tensor):
+            common.get_u55_compile_spec(permute_memory_to_nhwc=True),
+            test_data,
+        )
+        if common.is_option_enabled("corstone300"):
```
how would this run u85 bitstream?

per replied:
The compilespec is for Ethos-U55 and the run_method_and_compare_outpus is only run here for.

digantdesai replied:
Right, but this function can be called for both u55 and u85 IIUC

digantdesai commented on backends/arm/test/ops/test_linear.py:
```diff
@@ -172,9 +176,7 @@ def _test_linear_tosa_u55_BI_pipeline(
             .serialize()
         )
-
-        if common.is_option_enabled("corstone300"):
```
I am not sure if I am following the pattern where we enable/disable these. Is this operator specific? I would rather tie it to the ip i.e. something like this, hypothetically

```
if common.is_option_enabled("corstone300") and compile_spec.ip == "u55":
```

per replied:
the `common.is_option_enabled("corstone300")` only says if the FVP is installed on the target and can be run. 
Right now it's only added for a few operators, but due to the time it adds to CI we probably want to move away from enabling it for all operator unit tests all the time.

digantdesai replied:
IIUC this FVP is u55 specific while this function can also get u85 compile spec and vela/et can produce u85 binary. I was trying to highlight that the choice of fvp 300 vs 320 should be tied to the compile spec somehow.

digantdesai reviewed (commented):
Let's fix some small issues?

per commented on backends/arm/test/ops/test_avg_pool.py:
```diff
@@ -141,6 +145,22 @@ def test_avgpool2d_tosa_u55_BI(
+        )
+
+    @parameterized.expand(test_data_suite)
+    @unittest.expectedFailure
```
The comment on line 134/138 (# Expected to fail since ArmQuantizer cannot quantize a AvgPool2D layer) covers both tests.

per commented on backends/arm/test/ops/test_conv.py:
```diff
@@ -315,6 +315,24 @@ def _test_conv2d_u55_BI_pipeline(
+            .partition()
+            .check_count({"torch.ops.higher_order.executorch_call_delegate": 1})
+            .check_not(["executorch_exir_dialects_edge__ops_aten_convolution_default"])
+            .to_executorch()
```
We only serialize if the test should be running on FVP. Since it's time consuming to run them all (and on an additional coming target) we probably want to run on FVP for full models only, but not decided internally yet.

per commented on backends/arm/test/ops/test_depthwise_conv.py:
```diff
@@ -201,18 +200,37 @@ def _test_dw_conv2d_u55_BI_pipeline(
+        )
+
+    @parameterized.expand(testsuite)
+    @unittest.expectedFailure
```
This is wrong and should be fixed. Sorry about that.

zingo replied:
pul request updated

per commented on backends/arm/test/ops/test_expand.py:
```diff
@@ -104,6 +106,15 @@ def test_expand_tosa_BI(self, test_input, multiples):
     # Expected failure since tosa.TILE is unsupported by Vela.
     @parameterized.expand(Expand.test_parameters)
-    @unittest.expectedFailure
+    @unittest.expectedFailure  # TODO: MLBEDSW-9386
```
No, it's an internal bug that is in progress to be sorted out in Vela.

per commented on backends/arm/test/ops/test_conv.py:
```diff
@@ -315,6 +315,24 @@ def _test_conv2d_u55_BI_pipeline(
             .to_executorch()
         )
 
+    def _test_conv2d_u85_BI_pipeline(
```
Good point. Will fix.

zingo replied:
pull request updated

per committed:
Fix tests to address review comments.

Signed-off-by: Per Åstrand <per.astrand@arm.com>

Change-Id: Ibee7fa81517bf766a550b97be97e2e3dbaed3a6c

zingo committed:
Merge branch 'pytorch:main' into Add-all-relevant-testcases-for-Ethos-U85

zingo committed:
Merge branch 'pytorch:main' into Add-all-relevant-testcases-for-Ethos-U85

zingo committed:
Arm backend: Remove testing of dw_conv2d_3x3_1x3x256x256_gp3_st1 on Ethos-U55

This tests was remporary added but fails again,
lets remove it from the testing for now.

Signed-off-by: Zingo Andersen <zingo.andersen@arm.com>
Change-Id: I519a8b13e50e12c4093f1b8e4d80ae35bded23ff

zingo commented on backends/arm/test/ops/test_depthwise_conv.py:
```diff
@@ -130,11 +131,6 @@
-    ("3x3_1x3x256x256_gp3_st1", dw_conv2d_3x3_1x3x256x256_gp3_st1, True)
-)
-
 
```
This tests is causing problem in your CI dw_conv2d_3x3_1x3x256x256_gp3_st1 but works when we test it, lets remove it again (it was added back last week) for now and fix that in a separate patch after all is in.

zingo replied:
As a note the test running this was before this patch marked as @unittest.expectedFailure  but that line was removed below in this patch.

zingo committed:
Merge branch 'pytorch:main' into Add-all-relevant-testcases-for-Ethos-U85

digantdesai reviewed (approved):
Stamping to unblock.

zingo commented:
Just a try to rebase it on latest after Fredriks pull request got merged if you did this already on the "imported" please ignore the update.

zingo committed:
Merge branch 'main' into Add-all-relevant-testcases-for-Ethos-U85

zingo committed:
Arm Backend: fix merged typo in test_mobilenet_v2_arm.py

Signed-off-by: Zingo Andersen <zingo.andersen@arm.com>
Change-Id: I72baec5ab6cc85cd4f628ab2ea14f1c02c060b66
